codec/fuzz/fuzz_targets/codec_roundtrip.rs

```diff
@@ -0,0 +1,173 @@
+        }
+    };
+
+    assert_eq!(socket.ip(), decoded.ip());
+    assert_eq!(socket.port(), decoded.port());
+}
+
+fn roundtrip_bytes(input_data_bytes: Bytes) {
+    let input_len = input_data_bytes.len();
+    let encoded_bytes = input_data_bytes.encode();
+
+    // Decode with too long length
+    assert!(matches!(
+        Bytes::decode_cfg(encoded_bytes.clone(), &(0..input_len).into()),
+        Err(Error::InvalidLength(_))
+    ));
+
+    // Decode with too short length
+    assert!(matches!(
+        Bytes::decode_cfg(encoded_bytes.clone(), &(input_len + 1..).into()),
+        Err(Error::InvalidLength(_))
+    ));
+
+    // Decode with full length
+    let decoded_bytes = Bytes::decode_cfg(encoded_bytes, &(input_len..=input_len).into())
+        .expect("Decoding failed when it should have succeeded!!");
+
+    // Check matching
+    assert_eq!(input_data_bytes, decoded_bytes);
+}
+
+fn roundtrip_primitive<T, X>(v: T)
+where
+    X: std::default::Default,
+    T: Encode + Decode + PartialEq + DecodeExt<X> + std::fmt::Debug,
+{
+    let encoded = v.encode();
+    let decoded = T::decode(&mut &*encoded).expect("Failed to decode primitive!");
+    assert_eq!(v, decoded);
+}
+
+// NOTE: Separate float cases to handle NaN comparisons
+// TODO should combine these functions with better generics
+fn roundtrip_primitive_f32(v: f32) {
+    let encoded = v.encode();
+    let decoded: f32 = f32::decode(&mut &*encoded)
+        .expect("Failed to decode f32!")
+        .into();
+    if v.is_nan() && decoded.is_nan() {
+        // Ignore the NaN case
+        return;
+    }
+    assert_eq!(v, decoded);
+}
+
+fn roundtrip_primitive_f64(v: f64) {
+    let encoded = v.encode();
+    let decoded: f64 = f64::decode(&mut &*encoded)
+        .expect("Failed to decode f64!")
+        .into();
+    if v.is_nan() && decoded.is_nan() {
+        // Ignore the NaN case
+        return;
+    }
+    assert_eq!(v, decoded);
+}
+
+fn roundtrip_map<K, V>(map: &HashMap<K, V>, range_cfg: RangeCfg, k_cfg: K::Cfg, v_cfg: V::Cfg)
+where
+    K: Write + EncodeSize + Read + Clone + Ord + Hash + Eq + std::fmt::Debug + PartialEq,
+    V: Write + EncodeSize + Read + Clone + std::fmt::Debug + PartialEq,
+    HashMap<K, V>:
+        Read<Cfg = (RangeCfg, (K::Cfg, V::Cfg))> + std::fmt::Debug + PartialEq + Write + EncodeSize,
+{
+    let encoded = map.encode();
+    let config_tuple = (range_cfg, (k_cfg, v_cfg));
+    // TODO could also assert encoded size here with type info
+    let decoded =
+        HashMap::<K, V>::decode_cfg(encoded, &config_tuple).expect("Failed to decode map!");
+    assert_eq!(map, &decoded);
+}
+
+fn roundtrip_vec<T, X>(vec: Vec<T>)
+where
+    X: std::default::Default,
+    T: Encode + Decode + PartialEq + DecodeExt<X> + std::fmt::Debug,
+{
+    let input_len = vec.len();
+    let encoded = vec.encode();
+    let decoded =
+        Vec::<T>::decode_range(encoded, input_len..=input_len).expect("Failed to decode Vec<T>!");
+    assert_eq!(vec, decoded);
```

Comment:
nit: inconsistency with `roundtrip_bytes`: does not check too-short or too-long and uses `decode_range` instead of `decode_cfg`. whatever we settle on should probably just be consistent at least